collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py

```diff
@@ -228,18 +286,46 @@ def mask_ids(self, ids):
           output_mask: list of binary variables which indicate what tokens has
             been masked (to calculate the loss function for these tokens only)
         """
+
+        # Whole-word masking by default, as it gives better performance.
+        cand_indexes = []
+        for (i, id) in enumerate(ids):
+
+            if len(cand_indexes) >= 1 and not self.tokenizer.ids_to_tokens(
+                    [id])[0].startswith('\u2581'):
+                cand_indexes[-1].append(id)
+            else:
+                cand_indexes.append([id])
+
         masked_ids = []
         output_mask = []
-        for i in ids:
-            if random.random() < self.mask_probability:
-                output_mask.append(1)
+        for cand_index in cand_indexes:
+            if (random.random() < self.mask_probability) and \
+                  cand_index[0] != self.tokenizer.special_tokens["[CLS]"] and \
+                  cand_index[0] != self.tokenizer.special_tokens["[SEP]"]:
                 if random.random() < 0.8:
-                    masked_ids.append(self.tokenizer.special_tokens["[MASK]"])
+                    for cand_index_i in cand_index:
+                        output_mask.append(1)
+                        masked_ids.append(
+                            self.tokenizer.special_tokens["[MASK]"])
                 elif random.random() < 0.5:
-                    masked_ids.append(random.randrange(self.vocab_size))
+                    for cand_index_i in cand_index:
+                        output_mask.append(1)
+                        for _ in range(10):
```

Comment:
why these 10 attempts? can we do something like random_word = special_token and then sample until it isn't?

Reply:
again, imitated the original BERT implementation - it shouldn't go over 2 iterations most of the time, but 10 is there just to make sure. I'll put the comment in.
Certainly there may be more elegant and fast way to do this.
Using `while` loop like you said may be another way.